### PR TITLE
[Gradient Compression] Add error feedback to layerwise PowerSGD

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49451 [Gradient Compression] Warm-start of PowerSGD
* #49435 [Gradient Compression] Replace the assertions in PowerSGD comm hook by stream syncrhonization
* **#49418 [Gradient Compression] Add error feedback to layerwise PowerSGD**
* #49417 [Gradient Compression] Implement the original layerwise PowerSGD

Ad error feedback to the original implementation of PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25555538](https://our.internmc.facebook.com/intern/diff/D25555538/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -122,6 +122,33 @@ def powerSGD_hook(
...
+        # The buckets can be rebuilt during training.
+        # In this case, the error tensor shape will not be aligned with the input tensor,
+        # and the error will be re-initialized as zeros.
+        if (
+            bucket_index in state.error_dict
+            and state.error_dict[bucket_index].shape[0] == total_length
+        ):
+            input_tensor.add_(state.error_dict[bucket_index])
+        else:
+            logging.info(
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:140]: Nit: can we bump this to a `logging.debug()`? It seems more useful for gradient compression developers rather than users.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 2]
```diff
@@ -122,6 +122,33 @@ def powerSGD_hook(
     input_tensor = bucket.get_tensors()[0]
     device = input_tensor.device
     dtype = input_tensor.dtype
+
+    # Incorporate the error from the previous state into the gradients.
+    bucket_index = bucket.get_index()
+    input_tensor_cp = None
+    total_length = input_tensor.shape[0]
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:129]: nit: if these vars are only used if `state.use_error_feedback`, can we just define them inside the if block?

**Review by rohan-varma — APPROVED** [review]
LGTM, minor comments in line

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 3]
```diff
@@ -242,6 +269,10 @@ def decompress(fut):
         for p, q, tensor in zip(ps, qs, high_rank_tensors):
             torch.matmul(p, q.t(), out=tensor)
             assert not torch.any(torch.isnan(tensor))
+
+        if state.use_error_feedback:
+            # Memorize the local errors.
+            state.error_dict[bucket_index] = input_tensor_cp - input_tensor
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:275]: Isn't the error input_tensor_cp - tensor, where `tensor` is the average approximation of `input_tensor`?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:129]: These vars actually are also used in the later decompress function.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 1 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:140]: I think this log info is equally important as the one ["Reducer buckets have been rebuilt in this iteration"](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L672). Both of these logs appear very infrequently.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 3 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:275]: `input_tensor` is already the average approximation. For each approximation tensor M = PQ, since both P and Q are already synchronized after 2 rounds of allreduce, M will also be exactly the same, and no need for another allreduce.
